src/SAWScript/Proof.hs

```diff
@@ -1334,7 +1334,17 @@ psStats = _psStats
 -- forall x in ps1, exists y in ps2 where x == y
 propsSubset :: SharedContext -> [Prop] -> [Prop] -> IO Bool
 propsSubset sc ps1 ps2 =
-  and <$> sequence [ propsElem sc x ps2 | x <- ps1 ]
+  -- For each x, check if x exists in ps2 by checking term identity using stAppIndex
+  -- If the check succeeds, return True. Otherwise use propsElem to do the more expensive
+  -- convertibility check.
+  and <$> sequence [ if idSubset (unProp x) then pure True else propsElem sc x ps2 | x <- ps1 ]
+  where
+    ps2Ids = foldr (\x idents -> case (unProp x) of
```

Comment:
I'd expect this early check to be happening already inside of `scConvertibleEval` (called by `propsElem`), see https://github.com/GaloisInc/saw-script/blob/af4dc592831cfec8e9ca08040b7f4aaa0f09b4ce/saw-core/src/Verifier/SAW/SharedTerm.hs#L947 I'd expect `propsElem` to already be fast in cases where terms have the same index. Do you have any insight into why that's not the case?

Reply:
Thanks for the review! Say seq1 = [tm11, tm12, ...], and seq2 = [tm21, tm22, ...]. Suppose tm11.stAppIndex == tm22.stAppIndex. But because of the order, we fist check whether tm11 could be converted to tm21. This check unfortunately can be very expensive and the code does not reach the early return.

Reply:
:+1: